hardening: kernel config modularization

vunnyso opened the pull request:
* Created separate Kconfig file to enable virtualization on top of hardened baseline kernel.
* Using kernel merge_config.sh to merge kconfig fragment

<!--
    Copyright 2023 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

## Description of changes

<!--
Summary of the proposed changes in the PR description in your own words. For dependency updates, please link to the changelog.
-->

## Checklist for things done

<!-- Please check, [X], to all that applies. Leave [ ] if an item does not apply but you have considered the check list item. Note that all of these are not hard requirements. They serve information to reviewers. When you fill the checklist, you indicate to reviewers you appreciate their work. -->

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [x] Author has run `nix flake check --accept-flake-config` and it passes
- [x] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing
**This feature is only enabled for Lenovo X1.**

 *Tested following to work*
  nix build .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug
  nix build .#checks.x86_64-linux.module-test-hardened-generic-host-kernel
  nix build .#checks.x86_64-linux.module-test-hardened-lenovo-x1-guest-guivm-kernel
  nix build .#checks.x86_64-linux.module-test-hardened-pkvm-kernel
  nix develop .#devShells.x86_64-linux.kernel-x86
  [ghaf-kernel-x86-devshell:~/ghaf/linux-6.6.7]$ eval $checkPhase

**How to test**
By default kernel config modularization is disabled, its can be enabled using below configuration:
```
diff --git a/targets/lenovo-x1-carbon.nix b/targets/lenovo-x1-carbon.nix
index 35c35bd1c8..0850dea156 100644
--- a/targets/lenovo-x1-carbon.nix
+++ b/targets/lenovo-x1-carbon.nix
@@ -322,13 +322,13 @@
             ghaf = {
               hardware.definition = hwDefinition;
               # To enable guest hardening enable host hardening first
-              host.kernel.hardening.enable = false;
-              host.kernel.hardening.virtualization.enable = false;
-              host.kernel.hardening.networking.enable = false;
-              host.kernel.hardening.inputdevices.enable = false;
+              host.kernel.hardening.enable = lib.mkForce true;
+              host.kernel.hardening.virtualization.enable = lib.mkForce true;
+              host.kernel.hardening.networking.enable = lib.mkForce true;
+              host.kernel.hardening.inputdevices.enable = lib.mkForce true;
 
-              guest.kernel.hardening.enable = false;
-              guest.kernel.hardening.graphics.enable = false;
+              guest.kernel.hardening.enable = lib.mkForce true;
+              guest.kernel.hardening.graphics.enable = lib.mkForce true;
 
               host.kernel.hardening.hypervisor.enable = false;
 
@@ -573,8 +573,8 @@
     }
     ../modules/hardware/x86_64-generic/kernel/host
     {
-      ghaf.host.kernel.hardening.usb.enable = false;
-      ghaf.host.kernel.hardening.debug.enable = false;
+      ghaf.host.kernel.hardening.usb.enable = lib.mkForce true;
+      ghaf.host.kernel.hardening.debug.enable = lib.mkForce true;
     }
   ];
   releaseModules = [
```



## Known Issues
1. Scroll with two finger swipe is not working in this build but works in mainline (checked with themisto 141)
2. Clicking touch pad (using a bit more pressure so that the touch pad clicks) goes every now and then unnoticed by the OS of this build. In the mainline this works.
More details : https://github.com/tiiuae/ghaf/pull/439#issuecomment-1911884251

jenninikko commented on docs/src/architecture/hardening.md:
```diff
@@ -68,3 +68,13 @@ The host kernel runs on bare metal. The kernel is provided either with Linux ups
 The host kernel hardening is based on Linux `make tinyconfig`. The default `tinyconfig` fails to assertions on NixOS without modifications. Assertions are fixed in `ghaf_host_hardened_baseline` Linux configuration under Ghaf `modules/host/`.
```
in **the** `ghaf_host_hardened_baseline` configuration

vunnyso replied:
Updated as per review comment.

jenninikko commented on docs/src/architecture/hardening.md:
```diff
@@ -68,3 +68,13 @@ The host kernel runs on bare metal. The kernel is provided either with Linux ups
+
+### Virtualization Support
+
+The host Virtualization support will add required kernel config dependency to Ghaf baseline by which NixOS have virtualization enabled. It can be enabled using following flag `host.kernel_virtualization_hardening.enable` for Lenovo X1.
```
...**the** required kernel config...
...to **the** Ghaf baseline...
...NixOS **has**...
It can be enabled **with the** following...

vunnyso replied:
Updated as per review comment.

jenninikko commented on docs/src/architecture/hardening.md:
```diff
@@ -68,3 +68,13 @@ The host kernel runs on bare metal. The kernel is provided either with Linux ups
+
+### Networking Support
+
+The host Networking support will add required kernel config dependency to Ghaf baseline by which NixOS have networking enabled, It can be enabled using following flag `host.kernel_networking_hardening.enable` for Lenovo X1.
```
...**the** required kernel config...
...to **the** Ghaf baseline...
...NixOS **has**...
It can be enabled **with the** following...

vunnyso replied:
Updated as per review comment.

jenninikko commented on docs/src/architecture/hardening.md:
```diff
@@ -68,3 +68,13 @@ The host kernel runs on bare metal. The kernel is provided either with Linux ups
+
+The host Networking support will add required kernel config dependency to Ghaf baseline by which NixOS have networking enabled, It can be enabled using following flag `host.kernel_networking_hardening.enable` for Lenovo X1.
+
+To have Ghaf GUI we need to enable both Virtualization and Networking support. As of now, the kernel builds and can boot on Lenovo X1.
```
**To enable GUI, set** both Virtualization and Networking support.

vunnyso replied:
Updated as per review comment.

jenninikko reviewed (changes requested):
Hei, nice job! Just a few comments ^_^

vilvo reviewed (changes requested):
Thanks, nice development on enabling module hardened kernel. Functionally it works.
Testing reveals some kernel-hardening-checker issues:
```
[vilvo@carrie:~/ghaf]$ nix develop .#devShells.x86_64-linux.kernel-x86
[ghaf-kernel-devshell:~/ghaf/linux-6.6.5]$ scp root@ghaf-host:/proc/config.gz .
config.gz                                 100%   23KB   4.6MB/s   00:00
[ghaf-kernel-devshell:~/ghaf/linux-6.6.5]$ kernel-hardening-checker -c config.gz |grep FAIL
CONFIG_SYN_COOKIES                      |kconfig|     y      |defconfig | self_protection  | FAIL: "is not set"
CONFIG_CFI_CLANG                        |kconfig|     y      |   kspp   | self_protection  | FAIL: is not found
CONFIG_CFI_PERMISSIVE                   |kconfig| is not set |   kspp   | self_protection  | FAIL: CONFIG_CFI_CLANG is not "y"
CONFIG_RANDOM_KMALLOC_CACHES            |kconfig|     y      |    my    | self_protection  | FAIL: "is not set"
CONFIG_SECURITY_SELINUX                 |kconfig|     y      |    my    | security_policy  | FAIL: "is not set"
CONFIG_INET_DIAG                        |kconfig| is not set |   kspp   |cut_attack_surface| FAIL: "y"
CONFIG_MODULES                          |kconfig| is not set |   kspp   |cut_attack_surface| FAIL: "y"
CONFIG_FAIL_FUTEX                       |kconfig| is not set |  grsec   |cut_attack_surface| OK: is not found
CONFIG_FB                               |kconfig| is not set |maintainer|cut_attack_surface| FAIL: "y"
CONFIG_VT                               |kconfig| is not set |maintainer|cut_attack_surface| FAIL: "y"
CONFIG_KSM                              |kconfig| is not set |  clipos  |cut_attack_surface| FAIL: "y"
CONFIG_AIO                              |kconfig| is not set |  clipos  |cut_attack_surface| FAIL: "y"
CONFIG_TRIM_UNUSED_KSYMS                |kconfig|     y      |    my    |cut_attack_surface| FAIL: "is not set"
[+] Config check is finished: 'OK' - 184 / 'FAIL' - 12
[ghaf-kernel-devshell:~/ghaf/linux-6.6.5]$
```

I propose following module interface and structural changes - please see commit message for more detailed rationale: https://github.com/vilvo/ghaf/commit/b58ae30c867e7ba6b9d474de71048b7f38c6faa7

Finally, I think we should add `modules/host/kernel/test/default.nix` which enables us to build the kernel with hardening options even if it's not yet enabled for targets like Lenovo X1 by default. Test would allow us build the hardened kernel package in the CI and also run `kernel-hardening-checker` for any possible regression. Ping @avnik

vilvo commented:
@jenninikko @vunnyso - please "Hide" your comments if/when resolved. 

To hide resolved comment, please
1. click the three dots in your comment:
![screencapture](https://github.com/tiiuae/ghaf/assets/1027150/6d28df89-e9f9-482e-989b-1618c7fb7d02)
2. select hide
![screencapture](https://github.com/tiiuae/ghaf/assets/1027150/25ec08b4-8d57-4410-98dc-4f4e381f0184)
3. choose reason to hide the comment
![screencapture](https://github.com/tiiuae/ghaf/assets/1027150/98c17d91-12f2-4b43-b1c5-f7c0d507485f)

It will make the review comment resolution much easier to follow. Thanks!

vilvo commented:
@barnabakos - Thanks for testing already. Just FYI, as @vunnyso did not describe the testing procedure, the hardened kernel is not built *unless* you set these values `true` - @vunnyso please describe the testing procedure in the PR description "Testing"-section.

https://github.com/tiiuae/ghaf/pull/439/files#diff-789892fdf4e4c650d81ce42dbc84586807ced5e73a8d53c06775acf7df2d0876R212-R214 - also, the options support only Lenovo X1 Carbon at the moment so no need to test on NVIDIA Orin AGX. Kernel hardening for NVIDIA Orgin AGX is future work, out of scope for this PR.

vunnyso commented:
> Thanks, nice development on enabling module hardened kernel. Functionally it works. Testing reveals some kernel-hardening-checker issues:
> 
> ```
> [vilvo@carrie:~/ghaf]$ nix develop .#devShells.x86_64-linux.kernel-x86
> [ghaf-kernel-devshell:~/ghaf/linux-6.6.5]$ scp root@ghaf-host:/proc/config.gz .
> config.gz                                 100%   23KB   4.6MB/s   00:00
> [ghaf-kernel-devshell:~/ghaf/linux-6.6.5]$ kernel-hardening-checker -c config.gz |grep FAIL
> CONFIG_SYN_COOKIES                      |kconfig|     y      |defconfig | self_protection  | FAIL: "is not set"
> CONFIG_CFI_CLANG                        |kconfig|     y      |   kspp   | self_protection  | FAIL: is not found
> CONFIG_CFI_PERMISSIVE                   |kconfig| is not set |   kspp   | self_protection  | FAIL: CONFIG_CFI_CLANG is not "y"
> CONFIG_RANDOM_KMALLOC_CACHES            |kconfig|     y      |    my    | self_protection  | FAIL: "is not set"
> CONFIG_SECURITY_SELINUX                 |kconfig|     y      |    my    | security_policy  | FAIL: "is not set"
> CONFIG_INET_DIAG                        |kconfig| is not set |   kspp   |cut_attack_surface| FAIL: "y"
> CONFIG_MODULES                          |kconfig| is not set |   kspp   |cut_attack_surface| FAIL: "y"
> CONFIG_FAIL_FUTEX                       |kconfig| is not set |  grsec   |cut_attack_surface| OK: is not found
> CONFIG_FB                               |kconfig| is not set |maintainer|cut_attack_surface| FAIL: "y"
> CONFIG_VT                               |kconfig| is not set |maintainer|cut_attack_surface| FAIL: "y"
> CONFIG_KSM                              |kconfig| is not set |  clipos  |cut_attack_surface| FAIL: "y"
> CONFIG_AIO                              |kconfig| is not set |  clipos  |cut_attack_surface| FAIL: "y"
> CONFIG_TRIM_UNUSED_KSYMS                |kconfig|     y      |    my    |cut_attack_surface| FAIL: "is not set"
> [+] Config check is finished: 'OK' - 184 / 'FAIL' - 12
> [ghaf-kernel-devshell:~/ghaf/linux-6.6.5]$
> ```
> 
> I propose following module interface and structural changes - please see commit message for more detailed rationale: [vilvo@b58ae30](https://github.com/vilvo/ghaf/commit/b58ae30c867e7ba6b9d474de71048b7f38c6faa7)
> 
> Finally, I think we should add `modules/host/kernel/test/default.nix` which enables us to build the kernel with hardening options even if it's not yet enabled for targets like Lenovo X1 by default. Test would allow us build the hardened kernel package in the CI and also run `kernel-hardening-checker` for any possible regression. Ping @avnik

Sure @vilvo  that's good idea. Can I do `modules/host/kernel/test/default.nix` test addition as separate PR?

vunnyso commented:
> @barnabakos - Thanks for testing already. Just FYI, as @vunnyso did not describe the testing procedure, the hardened kernel is not built _unless_ you set these values `true` - @vunnyso please describe the testing procedure in the PR description "Testing"-section.
> 
> https://github.com/tiiuae/ghaf/pull/439/files#diff-789892fdf4e4c650d81ce42dbc84586807ced5e73a8d53c06775acf7df2d0876R212-R214 - also, the options support only Lenovo X1 Carbon at the moment so no need to test on NVIDIA Orin AGX. Kernel hardening for NVIDIA Orgin AGX is future work, out of scope for this PR.

@barnabakos and @vilvo I have updated "Testing" section in PR description.

vunnyso commented:
Added usb support commit, ready for review again.

vilvo commented:
> Added usb support commit, ready for review again.

I don't see the module interface change I proposed https://github.com/vilvo/ghaf/commit/b58ae30c867e7ba6b9d474de71048b7f38c6faa7
accounted in the latest changes.

Did you plan to:
- include it in the PR?
- ignore it for later?

I still think the module interface in this PR should be improved.

> Sure @vilvo that's good idea. Can I do `modules/host/kernel/test/default.nix` test addition as separate PR?

I drafted a module test yesterday and this morning - https://github.com/vilvo/ghaf/commit/ddcdd06723c8fcb9a4426a4eb1fe1b541ca56337
It's not final yet but already runs the host kernel build with hardened configuration and integrates to `nix run .#checks`
In general, review/merge of this PR is not in a hurry as this does not make the hardened kernel configuration default yet. So we can improve it and remove blockers to make the hardened kernel default.

vunnyso commented:
> > Added usb support commit, ready for review again.
> 
> I don't see the module interface change I proposed [vilvo@b58ae30](https://github.com/vilvo/ghaf/commit/b58ae30c867e7ba6b9d474de71048b7f38c6faa7) accounted in the latest changes.
> 
> Did you plan to:
> 
> * include it in the PR?
> * ignore it for later?
> 
> I still think the module interface in this PR should be improved.
> 
> > Sure @vilvo that's good idea. Can I do `modules/host/kernel/test/default.nix` test addition as separate PR?
> 
> I drafted a module test yesterday and this morning - [vilvo@ddcdd06](https://github.com/vilvo/ghaf/commit/ddcdd06723c8fcb9a4426a4eb1fe1b541ca56337) It's not final yet but already runs the host kernel build with hardened configuration and integrates to `nix run .#checks` In general, review/merge of this PR is not in a hurry as this does not make the hardened kernel configuration default yet. So we can improve it and remove blockers to make the hardened kernel default.

I have squashed  https://github.com/vilvo/ghaf/commit/b58ae30c867e7ba6b9d474de71048b7f38c6faa7 changes into c7f191c68774963215a4edf759a0b1c83a770197 

Noted on draft changes and timeline as well.

mikatammi commented on modules/host/kernel/buildKernel.nix:
```diff
@@ -0,0 +1,102 @@
+      base_kernel.overrideAttrs (_old: {
+        postUnpack = ''
+          mkdir $dev
+           cp $sourceRoot/scripts/kconfig/merge_config.sh $dev;
```
Indentation

vunnyso replied:
postUnpack is not needed anymore.

mikatammi commented on modules/host/kernel/buildKernel.nix:
```diff
@@ -0,0 +1,102 @@
+
+        inherit kernel_features;
+        postConfigure = ''
+          $dev/merge_config.sh  -O $buildRoot $buildRoot/.config  $kernel_features;
```
Do you actually need to copy this script to dev-output to be able to run it? It feels wrong

mikatammi replied:
From this point it should be available by just calling `scripts/kconfig/merge_config.sh` I think

vunnyso replied:
Yes you are right, updated code accordingly.

mikatammi commented on overlays/custom-packages/kernel/default.nix:
```diff
@@ -0,0 +1,6 @@
+# Copyright 2022-2024 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: Apache-2.0
+(_final: prev: {
+  makeModulesClosure = x:
```
Could this stuff only be enabled when this kernel config stuff is enabled? I mean, this stuff could be moved to `modules/host/kernel/default.nix`

vunnyso replied:
Updated as per review comment.

mikatammi commented on modules/host/kernel/default.nix:
```diff
@@ -0,0 +1,88 @@
+      }
+    else pkgs.linux_latest;
+
+  buildKernel = pkgs.callPackage ./buildKernel.nix {};
```
Using callPackage here feels wrong for some reason

mikatammi replied:
or at least the way how it is used, it's only used to pass `pkgs` and `lib´

vunnyso commented:
> Co-authored-by: Ville Ilvonen <ville.ilvonen@unikie.com>



> > I have squashed [vilvo@b58ae30](https://github.com/vilvo/ghaf/commit/b58ae30c867e7ba6b9d474de71048b7f38c6faa7) changes into [c7f191c](https://github.com/tiiuae/ghaf/commit/c7f191c68774963215a4edf759a0b1c83a770197)
> 
> Ah, my bad. I had looked at an older set of commits.
> 
> Now that only one of the commits is signed (Verified) and two not, it would be better to squash the commits into one and sign it. Please change the commit message end as follows:
> 
> ```
> Author: Vunny Sodhi <vunny.sodhi@unikie.com>
> 
> <combined commit message here>
> 
> Co-authored-by: Ville Ilvonen <ville.ilvonen@unikie.com>
> 
> Signed-off-by: Ville Ilvonen <ville.ilvonen@unikie.com>
> Signed-off-by: Vunny Sodhi <vunny.sodhi@unikie.com>
> ```
> 
> [More info on co-authored commits.](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors)
> 
> > Noted on draft changes and timeline as well.
> 
> Ok, let's continue finishing the module test for the module then before next round of reviews. Nice work so far!

Squashed commits, set to verified and added Co-Author.

vilvo commented on docs/src/architecture/hardening.md:
```diff
@@ -65,6 +68,28 @@ The host kernel runs on bare metal. The kernel is provided either with Linux ups
+
+The host USB support will add the required kernel config dependency to the Ghaf baseline by which NixOS has USB enabled, It can be enabled with the following flag `ghaf.host.kernel_usb_hardening.enable` for Lenovo X1.
+
+USB support needs to be enabled in case of external SSD and is optional in case of interval NVME.
```
Proposing:
USB support on host is for the `-debug`-profile builds, not for hardened host `-release`-builds. As of now, USB support needs to be enabled when debug support to host via USB ethernet adapter (Lenovo X1) is needed or when the user want to boot Ghaf using an external SSD. It is optional in case Ghaf is used with internal NVME.

vunnyso replied:
Updated documentation.

vilvo commented on docs/src/architecture/hardening.md:
```diff
@@ -65,6 +68,28 @@ The host kernel runs on bare metal. The kernel is provided either with Linux ups
+
+### Guest Support
+
+The Guest support will add the required kernel config dependency to the Ghaf baseline by which NixOS has guest enabled, It can be enabled with the following flag `guest.hardening.enable` for Lenovo X1.
```
Proposing:
"The Guest support will add the required kernel config dependency to the Ghaf baseline by which NixOS has guest enabled. The added functionality is `vsock` for host-to-guest and guest-to-guest communication.

It can be enabled with the following flag `guest.hardening.enable` for Lenovo X1."

vunnyso replied:
Updated documentation.

vilvo commented on modules/host/kernel/guest.config:
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: CC-BY-SA-4.0
+#Below configurations needed for microvm kernel to boot
+CONFIG_SERIAL_8250=y
+CONFIG_SERIAL_8250_CONSOLE=y
```
Where are `SERIAL_8250` needed? `-debug`-profiles? Or does something break if those are left out?

vunnyso replied:
If I don't include SERIAL_8250 then I see below error in journalctl
**ghaf-host microvm@kernel-vm[2140]: 2024/01/24 13:37:01 socat[2140] E UNIX-CLIENT: kernel-vm.sock: No such file or directory**
And microvm kernel won't boot at all.

vilvo replied:
This is probably related to earlier fixed serial console support https://github.com/astro/microvm.nix/commit/ce0d2e393216e708d2abc228c68a7a8c2147aeec that @jpruiz84 got fixed as optional in microvm.nix just yesterday.

Let's keep it as open issue. For now, I'd say this is debug-feature and could be hardened if disabling it in microvm.nix declarations allows us to disable it in kernel config.

jpruiz84 replied:
Both are needed

CONFIG_SERIAL_8250=y
CONFIG_SERIAL_8250_CONSOLE=y

The Robot Framework checks if the netvm started correctly by checking the output of the serial console. Then, these modules are needed.

vilvo replied:
Thanks. But they could be `-debug`-profile specific to support that testing requirement? 

In other words, there should be nothing blocking microvm.nix declared `netvm` from starting when serial is disabled and removed from kernel configs with `-release`-build where we won't enable that test automation interface?

jpruiz84 replied:
It is not recommended to test netvm boot process only in the debug profile. What we should do is to change the test procedure to make sure that netvm starts in debug and release profiles. I propose to change the test case ["Wait until NetVM service started" ](https://github.com/tiiuae/ci-test-automation/blob/ed45b93b8f42d338bccaec872ac2a860d4c4e1c6/Robot-Framework/resources/ssh_keywords.resource#L182)
from checking the systemctl status microvm@netvm output to check with ssh connection. @azbeleva

vilvo commented on modules/host/kernel/networking.config:
```diff
@@ -0,0 +1,28 @@
+CONFIG_SYN_COOKIES=y
+CONFIG_NETWORK_SECMARK=y
+CONFIG_AUDIT=y
+CONFIG_SECURITY_SELINUX=y
```
I think `SELINUX` should go to baseline or optionally to a new access control fragment. It's also related to network but it's not network. Or how do you see this?

vunnyso replied:
The reason not adding `SELINUX` in baseline is because of it's dependency on SECURITY_NETWORK, AUDIT, NET and INET all of them also need to be added in baseline as well. 
![image](https://github.com/tiiuae/ghaf/assets/147831196/acb8d8f5-03fc-4bf2-a953-d4b9b3ae28ae)

Please suggest further.

vilvo replied:
This is ok. Thanks for clarification.

vilvo commented on modules/host/kernel/virtualization.config:
```diff
@@ -0,0 +1,31 @@
+CONFIG_INPUT_EVDEV=y
+CONFIG_I2C=y
+CONFIG_I2C_DESIGNWARE_PLATFORM=y
+CONFIG_I2C_HID_ACPI=y
```
What's the case with `I2C*` specifics? General or Lenovo X1 related?

vilvo replied:
Ok, as I suspected. These are `gui-vm`-specific and input-devices are not needed in other guests. They can also get very device (both input device and target device) specific depending on the target configuration.  

All-in-all, they are not needed in all the guest VMs. Can you please refactor these to `user-input-devices.config` and include the config fragment only in the GUI VM?

vunnyso replied:
Done refactoring as per review comment.

vilvo commented on modules/virtualization/microvm/kernelvm.nix:
Can you please rename this to `hardenedvm.nix`? Sure it's only kernel as of now but the idea is to have a template for `hardenedvm` and that will include also the user space hardening with sensible defaults.

vunnyso replied:
Sure, that make more sense. Done renaming.

vilvo reviewed (changes requested):
Please address minor clarification requests. Based on your answers, let's see if we want those to either `-debug`-supporting kernel config fragments or device specific (Lenovo X1) kernel config fragments. That could make it easier to have the baseline and functional fragments that are **generic to other devices** with or without `-debug`-profile.

vilvo reviewed (approved):
Lgtm. Imo the microvm.nix optional serial and input devices kernel config fragment can be scoped to future work.

vunnyso commented:
Added @leivos-unikie for review/testing on Lenovo X1.

leivos-unikie commented:
@vunnyso could you elaborate more what are the expected outcomes in testing with different hardening config options.

So far I have noticed that can't build if set this true
`              guest.hardening.enable = true;`

Also setting all hardening config options to true in modules/host/hardening.nix made build end up with error

```
warning: Git tree '/home/samuli/repos/vunnyso/ghaf' is dirty
error: getting status of '/nix/store/2rzzya8s6rippjxvm40f34n1ydpk5isj-source/modules/host/virtualization': No such file or directory
(use '--show-trace' to show detailed location information)
```

Setting only 
    ghaf.host.kernel_virtualization_hardening.enable = true;
or
    ghaf.host.kernel_networking_hardening.enable = true;
and other config options to false
didn't seem to have any effect on VMs or network operation.


Trying to build more option combinations...

vilvo commented:
> Trying to build more option combinations...

Please don't - that kind of option coverage testing is not needed in the context of this PR. 
As of now - only specific host kernel hardening and demo guest VM (with matching hardened kernel) enabling scenario is supported. We can also leave the demo hardened guest VM testing for now. 

With that, this work as I just tested it for something else:
```
diff --git a/targets/lenovo-x1-carbon.nix b/targets/lenovo-x1-carbon.nix
index 8bcfb96..3a05c6f 100644
--- a/targets/lenovo-x1-carbon.nix
+++ b/targets/lenovo-x1-carbon.nix
@@ -211,8 +211,8 @@

             ghaf = {
               hardware.definition = hwDefinition;
-              host.hardening.enable = false;
+              host.hardening.enable = true;

diff --git a/modules/host/hardening.nix b/modules/host/hardening.nix
index bc752c7..bd5dd85 100644
--- a/modules/host/hardening.nix
+++ b/modules/host/hardening.nix
@@ -14,11 +14,11 @@

   config = {
     # host kernel hardening
-    ghaf.host.kernel_baseline_hardening.enable = false;
-    ghaf.host.kernel_virtualization_hardening.enable = false;
-    ghaf.host.kernel_networking_hardening.enable = false;
-    ghaf.host.kernel_usb_hardening.enable = false;
-    ghaf.host.kernel_inputdevices_hardening.enable = false;
+    ghaf.host.kernel_baseline_hardening.enable = true;
+    ghaf.host.kernel_virtualization_hardening.enable = true;
+    ghaf.host.kernel_networking_hardening.enable = true;
+    ghaf.host.kernel_usb_hardening.enable = true;
+    ghaf.host.kernel_inputdevices_hardening.enable = true;
```
With that configuration, GUI-VM boots and you can run all the other app VMs (as is) but we'd like to know if there's any regression we have not noticed. 

We discussed that option coverage testing - nix module testing - is to be enabled in future work. I've got a draft commit on it but let's leave it aside for now.

leivos-unikie commented:
> > Trying to build more option combinations...
> 
> Please don't - that kind of option coverage testing is not needed in the context of this PR. As of now - only specific host kernel hardening and demo guest VM (with matching hardened kernel) enabling scenario is supported. We can also leave the demo hardened guest VM testing for now.
> 
> With that, this work as I just tested it for something else:
> 
> ```
> diff --git a/targets/lenovo-x1-carbon.nix b/targets/lenovo-x1-carbon.nix
> index 8bcfb96..3a05c6f 100644
> --- a/targets/lenovo-x1-carbon.nix
> +++ b/targets/lenovo-x1-carbon.nix
> @@ -211,8 +211,8 @@
> 
>              ghaf = {
>                hardware.definition = hwDefinition;
> -              host.hardening.enable = false;
> +              host.hardening.enable = true;
> 
> diff --git a/modules/host/hardening.nix b/modules/host/hardening.nix
> index bc752c7..bd5dd85 100644
> --- a/modules/host/hardening.nix
> +++ b/modules/host/hardening.nix
> @@ -14,11 +14,11 @@
> 
>    config = {
>      # host kernel hardening
> -    ghaf.host.kernel_baseline_hardening.enable = false;
> -    ghaf.host.kernel_virtualization_hardening.enable = false;
> -    ghaf.host.kernel_networking_hardening.enable = false;
> -    ghaf.host.kernel_usb_hardening.enable = false;
> -    ghaf.host.kernel_inputdevices_hardening.enable = false;
> +    ghaf.host.kernel_baseline_hardening.enable = true;
> +    ghaf.host.kernel_virtualization_hardening.enable = true;
> +    ghaf.host.kernel_networking_hardening.enable = true;
> +    ghaf.host.kernel_usb_hardening.enable = true;
> +    ghaf.host.kernel_inputdevices_hardening.enable = true;
> ```
> 
> With that configuration, GUI-VM boots and you can run all the other app VMs (as is) but we'd like to know if there's any regression we have not noticed.
> 
> We discussed that option coverage testing - nix module testing - is to be enabled in future work. I've got a draft commit on it but let's leave it aside for now.

Thanks for the guidance. I'll stick in that config.

leivos-unikie commented:
Here, some findings from testing.

Configs:

targets/lenovo-x1-carbon.nix
            ghaf = {
              hardware.definition = hwDefinition;
              host.hardening.enable = true;
              guest.hardening.enable = false;

modules/host/hardening.nix
  config = {
    # host kernel hardening
    ghaf.host.kernel_baseline_hardening.enable = true;
    ghaf.host.kernel_virtualization_hardening.enable = true;
    ghaf.host.kernel_networking_hardening.enable = true;
    ghaf.host.kernel_usb_hardening.enable = true;
    ghaf.host.kernel_inputdevices_hardening.enable = true;
    # host kernel hypervisor (KVM) hardening
    ghaf.host.hypervisor_hardening.enable = false;
    # other host hardening options - user space, etc. - to be defined later
  };


It seems that Lenovo-x1 touch pad gestures have changed compared to ghaf mainline.
1. Scroll with two finger swipe is not working in this build but works in mainline (checked with themisto 141)
2. In this build soft tap works well as a mouse click. On the other hand it does not work in mainline. 
3. Clicking touch pad (using a bit more pressure so that the touch pad clicks) goes every now and then unnoticed by the OS of this build. In the mainline this works.

Tests of the ci-test-automation pass.

There is similar >20% drop in 1-thread memory/cpu performance tests in ghaf-host and all VMs. Multi-thread tests don't show such decrease in performance.

(Last test runs 11, 12 are with this PR.)
![cpu_1-thread](https://github.com/tiiuae/ghaf/assets/129752236/432ab50a-36cc-4f7a-b719-0a98e1d15203)
![mem_rd_1-thread](https://github.com/tiiuae/ghaf/assets/129752236/1bed6434-9c4d-4f54-b347-a7eb12ae9e71)
![mem_wr_1-thread](https://github.com/tiiuae/ghaf/assets/129752236/5b907094-6f68-440c-b365-6b00b45002a0)

gangaram-tii commented on modules/host/kernel/buildKernel.nix:
```diff
@@ -0,0 +1,110 @@
+      - also see https://github.com/NixOS/nixpkgs/issues/109280
+        for the context >
+      */
+      configfile = ./ghaf_host_hardened_baseline;
```
It is not guaranteed that this config file will work for newer/older kernels.   Is it possible to use defConfig as tinyconfig and enable other features (64-bit support, TTY, printk, and initrd etc) in extra config? It will dynamically generate minimal config for the kernel version in configuration phase.

vilvo replied:
You are absolutely right. Config will likely break with newer and older kernels when we update nixpkgs. But our approach here is to take different way from NixOS nixpkgs kernel config builder style which accepts the new kernel config option defaults automatically. Our approach is to check if we can safely disable the new or outdated options, then disable them to keep kernel better hardened with updates.

gangaram-tii replied:
Thanks Vilvo for the insight. So, for Ghaf customers it is not recommended to execute 'nix flake update'. It may pull newer Linux kernel which may not work.

vilvo replied:
In such scenario, the Ghaf-based project would be recommended https://tiiuae.github.io/ghaf/ref_impl/ghaf-based-project.html and Ghaf maintainers can choose secure defaults for new kernel config settings when newer kernels are updated.

vilvo commented:
> Here, some findings from testing.
> ...
> It seems that Lenovo-x1 touch pad gestures have changed compared to ghaf mainline.
> 
>     1. Scroll with two finger swipe is not working in this build but works in mainline (checked with themisto 141)
>     2. In this build soft tap works well as a mouse click. On the other hand it does not work in mainline.
>     3. Clicking touch pad (using a bit more pressure so that the touch pad clicks) goes every now and then unnoticed by the OS of this build. In the mainline this works.

Good findings. I think these are not blockers functionally but warrant further analysis. These input devices that implement touch pad are passed through to GUI-VM in both mainline and this. There might be something in the journalctl logs of `gui-vm` we have not noticed (e.g. warnings, missing input device config option/firmware in the gui-vm, etc.) but this can be worked further in future PRs - in my opinion. I know input devices test automation can be challenging but would be good to get something repeatable to use in development as well.

> 
> Tests of the ci-test-automation pass.
> 
> There is similar >20% drop in 1-thread memory/cpu performance tests in ghaf-host and all VMs. Multi-thread tests don't show such decrease in performance.

Thanks. Good findings. I'm not surprised as this is the first time `kernel-hardening-checker` enabled secure options are performance tested. I don't know the details of tests nor which kernel config hardening options could explain this as all the hardening options are enabled at once. If we want to study that further, bisecting the kernel config options (enabling them one by one in an automated build-testing script), could be one way. 

Anyway, I would scope that work outside this PR for the following reasons:
- this PR is structural, baseline kernel configuration and hardening implementation. Main focus in testing need was functional.
- these features are not enabled by default (yet) so further analysis is possible if/when enabled by default

vunnyso commented:
Hi @barnabakos and @mikatammi please review the code changes.

gangaram-tii commented:
Tested the PR on Gen-10 with following change set, I could not see graphics.
[x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt)

Tomorrow I will get Gen 11 from someone, will update the status of the PR on Gen-11.

vunnyso commented:
> Tested the PR on Gen-10 with following change set, I could not see graphics. [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt)
> 
> Tomorrow I will get Gen 11 from someone, will update the status of the PR on Gen-11.

Hi @gangaram-tii , may I know the reason in your [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt) why you have commented "# "module_blacklist=i915" ?

gangaram-tii commented:
This PR is working fine on "Lenovo carbon X1 Gen11"

vunnyso commented:
> > > Tested the PR on Gen-10 with following change set, I could not see graphics. [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt)
> > > Tomorrow I will get Gen 11 from someone, will update the status of the PR on Gen-11.
> > 
> > 
> > Hi @gangaram-tii , may I know the reason in your [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt) why you have commented "# "module_blacklist=i915" ?
> 
> I compiled it for Gen-10.

I understood you have compiled for Gen-10, I am asking why Intel i915 display driver blacklisting is commented out.

gangaram-tii commented:
> > > > Tested the PR on Gen-10 with following change set, I could not see graphics. [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt)
> > > > Tomorrow I will get Gen 11 from someone, will update the status of the PR on Gen-11.
> > > 
> > > 
> > > Hi @gangaram-tii , may I know the reason in your [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt) why you have commented "# "module_blacklist=i915" ?
> > 
> > 
> > I compiled it for Gen-10.
> 
> I understood you have compiled for Gen-10, I am asking why Intel i915 display driver blacklisting is commented out.

I think i915 kernel module is required for Gen10, it should not be blacklisted. I got this patch from someone else to build X10 image from X11 target. You can refer this discussion on Lenovo forum:
https://forums.lenovo.com/t5/Other-Linux-Discussions/Problems-with-graphics-in-linux-on-Thinkpad-X1-Gen-10/m-p/5192427

It didn't work on Gen10 with updated flake, same issue came on Gen11 with updated flake. With original flake it is working fine.

vunnyso commented:
> > > > > Tested the PR on Gen-10 with following change set, I could not see graphics. [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt)
> > > > > Tomorrow I will get Gen 11 from someone, will update the status of the PR on Gen-11.
> > > > 
> > > > 
> > > > Hi @gangaram-tii , may I know the reason in your [x1_gen10.txt](https://github.com/tiiuae/ghaf/files/14111785/x1_gen10.txt) why you have commented "# "module_blacklist=i915" ?
> > > 
> > > 
> > > I compiled it for Gen-10.
> > 
> > 
> > I understood you have compiled for Gen-10, I am asking why Intel i915 display driver blacklisting is commented out.
> 
> I think i915 kernel module is required for Gen10, it should not be blacklisted. I got this patch from someone else to build X10 image from X11 target. You can refer this discussion on Lenovo forum: https://forums.lenovo.com/t5/Other-Linux-Discussions/Problems-with-graphics-in-linux-on-Thinkpad-X1-Gen-10/m-p/5192427
> 
> It didn't work on Gen10 with updated flake, same issue came on Gen11 with updated flake. With original flake it is working fine.

Thanks for the info @gangaram-tii

vilvo commented:
@vunnyso - configs refactoring requested by @brianmcgillion is basically about looking at each config fragment and asking following questions:

- Is this kernel configuration fragment generic to all `x86_64`-devices - both host and guest?
  - If yes, place it under `modules/hardware/x86_64-generic/kernel/configs/`
  - If not, move to next question
- Is this kernel configuration fragment generic to all `x86_64`-device hosts?
  - If yes, place it under `modules/hardware/x86_64-generic/kernel/host/configs/`
  - If not, move to next question
- Is this kernel configuration fragment specific to the X1 host hardened kernel (e.g. device known only to be available in Lenovo X1 Carbon Gen11) ?
    - If yes, place it under `modules/hardware/lenovo-x1/kernel/host/configs/`
    - If not, move to next question
- Is this kernel configuration fragment generic to all `x86_64`-device guest VMs?
    - If yes, place it under `modules/hardware/x86_64-generic/kernel/guest/configs/`
    - If not, move to next question
- Is this kernel configuration fragment specific to a  X1 guest hardened kernel (e.g. pass-through device known only to be available in Lenovo X1 Carbon Gen11)?
    - If yes, place is under `modules/hardware/lenovo-x1/kernel/guest/configs/`
    - If not, raise an issue to ask for more info 
    
I wrote the WIP refactoring commit with module test support in [15a051f](https://github.com/tiiuae/ghaf/pull/439/commits/15a051f7ae0f37ddc1a2ad5347a4d1bc51f84cf0). I hope it works as an example.

vilvo commented on modules/hardware/x86_64-generic/kernel/host/configs/networking.config:
```diff
@@ -0,0 +1,28 @@
+CONFIG_PACKET=y
+CONFIG_BRIDGE=y
+CONFIG_NETDEVICES=y
+CONFIG_USB_USBNET=y
```
@vunnyso I think `CONFIG_USB_USBNET` should go to common `debug.config` fragment. What do you think?

vunnyso replied:
We can move `CONFIG_USB_USBNET` to `debug.config` only concern is  `CONFIG_USB_USBNET`  depends on `Depends on: NETDEVICES [=y] && USB_NET_DRIVERS [=y] and further depends on NET [=y]`, due to strong dependency on `networking.config` I have currently kept it under same file.

vilvo replied:
That way the dependency is ok. If networking.config is left out the build and platform functionality will break anyway. But if debug.config is left out, it will harden the configuration.

vilvo commented on modules/hardware/x86_64-generic/kernel/host/configs/usb.config:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: CC-BY-SA-4.0
+CONFIG_USB_SUPPORT=y
+CONFIG_USB=y
+CONFIG_USB_STORAGE=y
```
We could place `CONFIG_USB_STORAGE` in `debug.config` fragment as well. 
I don't see how it would be needed on host in generic case for `-release`-mode.

vunnyso replied:
That case I think all `usb.config` can be moved only for `-debug`-mode assuming release builds be flashed on NVME.

vilvo commented on modules/hardware/x86_64-generic/kernel/host/pkvm.nix:
@vunnyso - do you mind adding a module test for this as well? I think it would help development and testing of @Kamarjam

vilvo commented on modules/virtualization/microvm/guivm.nix:
```diff
@@ -118,6 +118,14 @@
+  # Importing kernel builder function from packages and checking hardening options
+  buildKernel = import ../../../packages/kernel {inherit config pkgs lib;};
+  config_baseline = ../../hardware/x86_64-generic/kernel/configs/ghaf_host_hardened_baseline-x86;
+  # TODO: Find proper way to place configs, below are commented as its causing build errors
```
This should be set in the target configuration - similar to the module tests. I think you can remove these from here.

vilvo commented on modules/virtualization/microvm/hardenedvm.nix:
This can be removed when there's a mechanism to enable hardened gui-vm guest for Lenovo X1 Carbon target configuration.

vilvo commented on targets/lenovo-x1-carbon.nix:
```diff
@@ -270,7 +272,16 @@
             ghaf = {
               hardware.definition = hwDefinition;
-              host.kernel_hardening.enable = false;
+              # To enable guest hardening enable host hardening first
```
And these must still be `false` by default.

vilvo commented:
Beyond above comments, I tested following to work:
```
  nix build .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug
  nix build .#checks.x86_64-linux.module-test-hardened-generic-host-kernel
  nix build .#checks.x86_64-linux.module-test-hardened-lenovo-x1-guest-guivm-kernel
  nix develop .#devShells.x86_64-linux.kernel-x86
  [ghaf-kernel-x86-devshell:~/ghaf/linux-6.6.7]$ eval $checkPhase
```

vunnyso commented:
Hi @mikatammi and @barnabakos please test the code changes for Lenovo X1.

avnik commented on nix/devshell/kernel.nix:
```diff
@@ -44,7 +44,7 @@
         '';
         # use "eval $checkPhase" - see https://discourse.nixos.org/t/nix-develop-and-checkphase/25707
-        checkPhase = "cp ../modules/host/ghaf_host_hardened_baseline-${platform} ./.config && make -j$(nproc)";
+        checkPhase = "cp ../modules/hardware/x86_64-generic/kernel/configs/ghaf_host_hardened_baseline-${platform} ./.config && make -j$(nproc)";
```
Using of `${platform}` is wrong here, `x86_64-generic` already a platform, so here should be something like `../modules/${platform}/kernel/configs/ghaf_baseline`

vunnyso replied:
I have addressed review comment.

Mic92 commented on modules/hardware/x86_64-generic/kernel/host/default.nix:
```diff
@@ -0,0 +1,72 @@
+      default = false;
+    };
+
+    options.ghaf.host.kernel.virtualization_hardening.enable = lib.mkOption {
```
Just a high level comment. The way all these features are written is a bit misleading.
Just now it reads as if this option would enable hardening for virtualization (effectively reducing the attack surface). However what it does is enables support for virtualisation, networking, input, usb...), so increases the surface. I would rewrite this the option as follow:

```
options.ghaf.host.kernel.baseline_hardening.enable -> options.ghaf.host.kernel.hardening.enable
```

And than add all other options like this:

```
options.ghaf.host.kernel.virtualisation_hardening.enable -> options.ghaf.host.kernel.hardening.virtualisation.enable
```

And update the options description like this:

```
description = "Enable support for virtualization in the Ghaf Host";
```

vilvo replied:
I like the proposal. Literate programming makes the intent more explicit here. @vunnyso - do you mind rewriting the module options as proposed?
@Mic92 Not related to your previous comment, but this may help with the PR design comprehension https://ssrc.atlassian.net/wiki/spaces/GA/pages/934445346/Hardened+kernel+config+modularization

vunnyso replied:
Renamed variables accordingly.

Mic92 commented on modules/hardware/x86_64-generic/kernel/host/pkvm/default.nix:
```diff
@@ -0,0 +1,48 @@
+        url = "mirror://kernel/linux/kernel/v6.x/linux-${version}.tar.xz";
+        hash = "sha256-qH4kHsFdU0UsTv4hlxOjdp2IzENrW5jPbvsmLEr/FcA=";
+      };
+      version = "6.1.55";
+      modDirVersion = "6.1.55";
```
```suggestion
      inherit version;
      modDirVersion = version;
```

vunnyso replied:
Modified accordingly.

Mic92 commented on modules/hardware/x86_64-generic/kernel/host/pkvm/default.nix:
```diff
@@ -0,0 +1,48 @@
+    };
+    config = mkIf hyp_cfg.enable {
+      boot.kernelPackages = pkgs.linuxPackagesFor pkvmKernel;
+      boot.kernelPatches = mkIf (hyp_cfg.enable && "${pkvmKernel.version}" == "6.1.55") pkvm_patch;
```
```suggestion
      boot.kernelPatches = mkIf (hyp_cfg.enable && "${pkvmKernel.version}" == version) pkvm_patch;
```

Mic92 replied:
Not quite sure what the version check does here given that `kernelPackages` already will enforce a certain kernel version already?

Mic92 commented on modules/hardware/x86_64-generic/kernel/host/pkvm/default.nix:
```diff
@@ -0,0 +1,48 @@
+  lib,
+  pkgs,
+  ...
+}: let
```
The update script would be not a blocker for this PR for sure. Maybe another issue?

vilvo commented:
@Mic92 short info for the background on the `pkvm` kernel:
`modules/hardware/x86_64-generic/kernel/host/pkvm/*`
It's a study kernel on Google/Intel special kernel with KVM security extensions. Within the scope of this PR, we only refactored it to a package of its' own - separated from the default hardened kernel builder - and added a module test to build it because it is disabled by default. hardened kernel and `pkvm` kernel do not combine at the moment.
The pkvm-kernel for `x86_64` *is not updated* regularly. More info from @Kamarjam and I leave it to him to consider the proposed changes.
